README.md
# Transit.land API wrapper

Simple Python wrapper around <https://transit.land> API.

## Install

```
pip install transitland_wrapper
```

## CLI

All CLI commands write a list of GeoJSON `Feature`s to stdout. This way minimal
results are kept in memory at once. If a `geometry` is passed, the geometry's
bounding box is given to the transit.land API, and results are tested for
intersection with the original geometry. The `geometry` file must be readable by
GeoPandas and will automatically be reprojected to EPSG 4326 if necessary. The
geometry type must be either `Point`, `Polygon`, `MultiPolygon`, `LineString`,
or `MultiLineString`.

### Operators

```
Usage: transitland operators [OPTIONS]

  Request operators info

Options:
  -b, --bbox TEXT             Bounding box to search within
  -g, --geometry PATH         File with geometry to use. Must be readable by
                              GeoPandas
  -r, --radius FLOAT          radius in meters to search around, default 100m
                              for Point geometries. Used only for Point
                              geometries.
  --gtfs-id TEXT              ID used in a GTFS feed's agencies.txt file
  -p, --per-page INTEGER      Number of results per page  [default: 50]
  --page-all / --no-page-all  Page over all responses  [default: False]
  --help                      Show this message and exit.
```


### Routes

```
Usage: transitland routes [OPTIONS]

  Request routes info

Options:
  -b, --bbox TEXT                 Bounding box to search within
  -g, --geometry PATH             File with geometry to use. Must be readable
                                  by GeoPandas
  -r, --radius FLOAT              radius in meters to search around, default
                                  100m for Point geometries. Used only for
                                  Point geometries.
  --operated-by TEXT              search by operator onestop_id or route
                                  onestop_id
  --vehicle-type TEXT             find all routes with vehicle type(s) by
                                  integer or string. Possible values defined
                                  by the GTFS spec for the route_type column
                                  and the Extended GTFS Route Types
  --gtfs-id TEXT                  ID used in a GTFS feed's routes.txt file
  --include-geometry / --no-include-geometry
                                  Include route geometry  [default: True]
  -p, --per-page INTEGER          Number of results per page  [default: 50]
  --page-all / --no-page-all      Page over all responses  [default: False]
  --help                          Show this message and exit.
```

### Stops

```
Usage: transitland stops [OPTIONS]

  Request stops info

Options:
  -b, --bbox TEXT             Bounding box to search within
  -g, --geometry PATH         File with geometry to use. Must be readable by
                              GeoPandas
  -r, --radius FLOAT          radius in meters to search around, default 100m
                              for Point geometries. Used only for Point
                              geometries.
  --served-by TEXT            search by operator onestop_id or route
                              onestop_id
  --gtfs-id TEXT              ID used in a GTFS feed's stops.txt file
  -p, --per-page INTEGER      Number of results per page  [default: 50]
  --page-all / --no-page-all  Page over all responses  [default: False]
  --help                      Show this message and exit.
```

### Route Stop Patterns

```
Usage: transitland route-stop-patterns [OPTIONS]

  Request routes info

Options:
  -b, --bbox TEXT             Bounding box to search within
  -g, --geometry PATH         File with geometry to use. Must be readable by
                              GeoPandas
  --traversed-by TEXT         find all Route Stop Patterns belonging to route
  --stops-visited TEXT        any one or more stop Onestop IDs, separated by
                              comma. Finds Route Stop Patterns with
                              stops_visited in stop_pattern
  --trips TEXT                any one or more trip ids, separated by comma.
                              Finds Route Stop Patterns with specified trips
                              in trips
  -p, --per-page INTEGER      Number of results per page  [default: 50]
  --page-all / --no-page-all  Page over all responses  [default: False]
  --help                      Show this message and exit.
```

### Schedule Stop Pairs

```
Usage: transitland schedule-stop-pairs [OPTIONS]

  Request schedule stop pairs info

Options:
  -b, --bbox TEXT                 Bounding box to search within
  -g, --geometry PATH             File with geometry to use. Must be readable
                                  by GeoPandas
  --origin-onestop-id TEXT        Find all Schedule Stop Pairs from origin
  --destination-onestop-id TEXT   Find all Schedule Stop Pairs to a
                                  destination
  --date TEXT                     Find all Schedule Stop Pairs from origin on
                                  date
  --service-from-date TEXT        Find all Schedule Stop Pairs in effect from
                                  a date
  --service-before-date TEXT      Find all Schedule Stop Pairs in effect
                                  before a date
  --origin-departure-between TEXT
                                  Find all Schedule Stop Pairs with
                                  origin_departure_time in a range
  --trip TEXT                     Find all Schedule Stop Pairs by trip
                                  identifier
  --route-onestop-id TEXT         Find all Schedule Stop Pairs by route
  --operator-onestop-id TEXT      Find all Schedule Stop Pairs by operator
  --active / --no-active          Schedule Stop Pairs from active FeedVersions
                                  [default: True]
  -p, --per-page INTEGER          Number of results per page  [default: 50]
  --page-all / --no-page-all      Page over all responses  [default: False]
  --help                          Show this message and exit.
```

### Onestop ID

```
Usage: transitland onestop-id [OPTIONS]

  Request onestop_id info

Options:
  --oid TEXT       a Onestop ID for any type of entity (for example, a stop or
                   an operator)
  -f, --file PATH  a file with one or more Onestop IDs, with each on their own
                   line.
  --help           Show this message and exit.
```

### Feeds

```
Usage: transitland feeds [OPTIONS]

  Request feeds info

Options:
  -b, --bbox TEXT             Bounding box to search within
  -g, --geometry PATH         File with geometry to use. Must be readable by
                              GeoPandas
  -p, --per-page INTEGER      Number of results per page  [default: 50]
  --page-all / --no-page-all  Page over all responses  [default: False]
  --help                      Show this message and exit.
```

## Python API

Each function returns a _generator_ of results, because there could be an
unknown amount of paging involved. Each item of the iterator is a list of
GeoJSON `Feature`s.

```py
import transitland_wrapper
transitland_wrapper.stops()
transitland_wrapper.operators()
transitland_wrapper.routes()
transitland_wrapper.route_stop_patterns()
transitland_wrapper.schedule_stop_pairs()
transitland_wrapper.onestop_id()
transitland_wrapper.feeds()
```

### Stops

```
- geometry: either Point, to search a radius around a point, or a
  Polygon or MultiPolygon, to search for stops within the geometry. If a
  Polygon or MultiPolygon is provided, the search will be done by
  bounding box, and then results will be filtered for intersection.
- radius: radius in meters to search around, default 100m for Point
  geometries. Not used for Polygon geometries.
- served_by: search by operator onestop_id or route onestop_id
- gtfs_id: ID used in a GTFS feed's stops.txt file
- per_page: number of results per page, by default 50
- page_all: page over all responses
```

### Operators

```
- geometry: either Point, to search a radius around a point, or a
  Polygon or MultiPolygon, to search for stops within the geometry. If a
  Polygon or MultiPolygon is provided, the search will be done by
  bounding box, and then results will be filtered for intersection.
- radius: radius in meters to search around, default 100m for Point
  geometries. Not used for Polygon geometries.
- gtfs_id: ID used in a GTFS feed's agencies.txt file
- per_page: number of results per page, by default 50
- page_all: page over all responses
```

### Routes

```
- geometry: either Point, to search a radius around a point, or a
  Polygon or MultiPolygon, to search for stops within the geometry. If a
  Polygon or MultiPolygon is provided, the search will be done by
  bounding box, and then results will be filtered for intersection.
- radius: radius in meters to search around, default 100m for Point
  geometries. Not used for Polygon geometries.
- operated_by: search by operator onestop_id
- vehicle_type: find all routes with vehicle type(s) by integer or
  string. Possible values defined by the GTFS spec for the route_type
  column and the Extended GTFS Route Types.
- include_geometry: If True, includes route geometry. Default: True
- gtfs_id: ID used in a GTFS feed's routes.txt file
- per_page: number of results per page, by default 50
- page_all: page over all responses
```

### Route Stop Patterns

```
- geometry: either Polygon or MultiPolygon, to search for stops within
  the geometry. If a Polygon or MultiPolygon is provided, the search
  will be done by bounding box, and then results will be filtered for
  intersection.
- traversed_by: find all Route Stop Patterns belonging to route
- stops_visited: any one or more stop Onestop IDs, separated by comma. Finds Route Stop Patterns with stops_visited in stop_pattern.
- trips: any one or more trip ids, separated by comma. Finds Route Stop Patterns with specified trips in trips.
- per_page: number of results per page, by default 50
- page_all: page over all responses
```

### Schedule Stop Pairs

```
- geometry: either Polygon or MultiPolygon, to search for stops within
  the geometry. If a Polygon or MultiPolygon is provided, the search
  will be done by bounding box, and then results will be filtered for
  intersection.
- origin_onestop_id: Find all Schedule Stop Pairs from origin. Accepts multiple Onestop IDs, separated by commas
- destination_onestop_id: Find all Schedule Stop Pairs to a destination. Accepts multiple Onestop IDs, separated by commas
- date: Find all Schedule Stop Pairs from origin on date
- service_from_date: Find all Schedule Stop Pairs in effect from a date
- service_before_date: Find all Schedule Stop Pairs in effect before a date
- origin_departure_between: Find all Schedule Stop Pairs with origin_departure_time in a range
- trip: Find all Schedule Stop Pairs by trip identifier
- route_onestop_id: Find all Schedule Stop Pairs by route. Accepts multiple Onestop IDs, separated by commas.
- operator_onestop_id: Find all Schedule Stop Pairs by operator. Accepts multiple Onestop IDs, separated by commas.
- active: Schedule Stop Pairs from active FeedVersions
- per_page: number of results per page, by default 50
- page_all: page over all responses
```

### Onestop ID

```
- oid: a Onestop ID for any type of entity (for example, a stop or an operator)
```

### Feeds

```
- geometry: either Polygon or MultiPolygon, to search for stops within
  the geometry. If a Polygon or MultiPolygon is provided, the search
  will be done by bounding box, and then results will be filtered for
  intersection.
```

## Contributing

To release to PyPI:
```
bumpversion minor
python setup.py sdist
twine upload dist/transitland_wrapper-0.1.0.tar.gz
```


CHANGELOG.md
# Changelog

## [0.5.0] - 2020-02-23

- Add feeds endpoint
- Set `page-all` to false by default

## [0.4.2] - 2020-02-19

- Fix bug where a string was treated as an iterable to be joined

## [0.4.1] - 2020-02-19

- Fix bug where the routes cli used the operators API endpoint

## [0.4.0] - 2020-02-09

- Add argument for paging size
- Add option to not page over all responses

## [0.3.0] - 2020-02-09

- Add onestop_id endpoint

## [0.2.0] - 2020-02-08

- Add route stop patterns and schedule stop pairs endpoints

## [0.1.1] - 2020-02-08

- Try to fix PyPI bundle

## [0.1.0] - 2020-02-08

- Initial release on PyPI


transitland_wrapper/cli.py
import json

import click
from shapely.geometry import box

from . import transitland


@click.group()
def main():
    pass


@click.command()
@click.option(
    '-b',
    '--bbox',
    required=False,
    default=None,
    type=str,
    help='Bounding box to search within')
@click.option(
    '-g',
    '--geometry',
    required=False,
    default=None,
    type=click.Path(exists=True, file_okay=True, readable=True),
    help='File with geometry to use. Must be readable by GeoPandas')
@click.option(
    '-r',
    '--radius',
    required=False,
    default=None,
    type=float,
    help=
    'radius in meters to search around, default 100m for Point geometries. Used only for Point geometries.'
)
@click.option(
    '--served-by',
    required=False,
    default=None,
    multiple=True,
    type=str,
    help='search by operator onestop_id or route onestop_id')
@click.option(
    '--gtfs-id',
    required=False,
    default=None,
    type=str,
    help="ID used in a GTFS feed's stops.txt file")
@click.option(
    '-p',
    '--per-page',
    required=False,
    default=50,
    show_default=True,
    type=int,
    help='Number of results per page')
@click.option(
    '--page-all/--no-page-all',
    is_flag=True,
    default=False,
    show_default=True,
    help='Page over all responses')
def stops(**kwargs):
    """Request stops info"""
    kwargs = handle_geometry(**kwargs)
    features_iter = transitland.stops(**kwargs)
    write_to_stdout(features_iter)


@click.command()
@click.option(
    '-b',
    '--bbox',
    required=False,
    default=None,
    type=str,
    help='Bounding box to search within')
@click.option(
    '-g',
    '--geometry',
    required=False,
    default=None,
    type=click.Path(exists=True, file_okay=True, readable=True),
    help='File with geometry to use. Must be readable by GeoPandas')
@click.option(
    '-r',
    '--radius',
    required=False,
    default=None,
    type=float,
    help=
    'radius in meters to search around, default 100m for Point geometries. Used only for Point geometries.'
)
@click.option(
    '--gtfs-id',
    required=False,
    default=None,
    type=str,
    help="ID used in a GTFS feed's agencies.txt file")
@click.option(
    '-p',
    '--per-page',
    required=False,
    default=50,
    show_default=True,
    type=int,
    help='Number of results per page')
@click.option(
    '--page-all/--no-page-all',
    is_flag=True,
    default=False,
    show_default=True,
    help='Page over all responses')
def operators(**kwargs):
    """Request operators info"""
    kwargs = handle_geometry(**kwargs)
    features_iter = transitland.operators(**kwargs)
    write_to_stdout(features_iter)


@click.command()
@click.option(
    '-b',
    '--bbox',
    required=False,
    default=None,
    type=str,
    help='Bounding box to search within')
@click.option(
    '-g',
    '--geometry',
    required=False,
    default=None,
    type=click.Path(exists=True, file_okay=True, readable=True),
    help='File with geometry to use. Must be readable by GeoPandas')
@click.option(
    '-r',
    '--radius',
    required=False,
    default=None,
    type=float,
    help=
    'radius in meters to search around, default 100m for Point geometries. Used only for Point geometries.'
)
@click.option(
    '--operated-by',
    required=False,
    default=None,
    type=str,
    help='search by operator onestop_id or route onestop_id')
@click.option(
    '--vehicle-type',
    required=False,
    default=None,
    multiple=True,
    type=str,
    help=
    'find all routes with vehicle type(s) by integer or string. Possible values defined by the GTFS spec for the route_type column and the Extended GTFS Route Types'
)
@click.option(
    '--gtfs-id',
    required=False,
    default=None,
    type=str,
    help="ID used in a GTFS feed's routes.txt file")
@click.option(
    '--include-geometry/--no-include-geometry',
    is_flag=True,
    default=True,
    show_default=True,
    help="Include route geometry")
@click.option(
    '-p',
    '--per-page',
    required=False,
    default=50,
    show_default=True,
    type=int,
    help='Number of results per page')
@click.option(
    '--page-all/--no-page-all',
    is_flag=True,
    default=False,
    show_default=True,
    help='Page over all responses')
def routes(**kwargs):
    """Request routes info"""
    kwargs = handle_geometry(**kwargs)
    features_iter = transitland.routes(**kwargs)
    write_to_stdout(features_iter)


@click.command()
@click.option(
    '-b',
    '--bbox',
    required=False,
    default=None,
    type=str,
    help='Bounding box to search within')
@click.option(
    '-g',
    '--geometry',
    required=False,
    default=None,
    type=click.Path(exists=True, file_okay=True, readable=True),
    help='File with geometry to use. Must be readable by GeoPandas')
@click.option(
    '--traversed-by',
    required=False,
    default=None,
    type=str,
    help='find all Route Stop Patterns belonging to route')
@click.option(
    '--stops-visited',
    required=False,
    default=None,
    multiple=True,
    type=str,
    help=
    'any one or more stop Onestop IDs, separated by comma. Finds Route Stop Patterns with stops_visited in stop_pattern'
)
@click.option(
    '--trips',
    required=False,
    default=None,
    multiple=True,
    type=str,
    help=
    'any one or more trip ids, separated by comma. Finds Route Stop Patterns with specified trips in trips'
)
@click.option(
    '-p',
    '--per-page',
    required=False,
    default=50,
    show_default=True,
    type=int,
    help='Number of results per page')
@click.option(
    '--page-all/--no-page-all',
    is_flag=True,
    default=False,
    show_default=True,
    help='Page over all responses')
def route_stop_patterns(**kwargs):
    """Request routes info"""
    kwargs = handle_geometry(**kwargs)
    features_iter = transitland.route_stop_patterns(**kwargs)
    write_to_stdout(features_iter)


@click.command()
@click.option(
    '-b',
    '--bbox',
    required=False,
    default=None,
    type=str,
    help='Bounding box to search within')
@click.option(
    '-g',
    '--geometry',
    required=False,
    default=None,
    type=click.Path(exists=True, file_okay=True, readable=True),
    help='File with geometry to use. Must be readable by GeoPandas')
@click.option(
    '--origin-onestop-id',
    required=False,
    multiple=True,
    default=None,
    type=str,
    help='Find all Schedule Stop Pairs from origin')
@click.option(
    '--destination-onestop-id',
    required=False,
    multiple=True,
    default=None,
    type=str,
    help='Find all Schedule Stop Pairs to a destination')
@click.option(
    '--date',
    required=False,
    default=None,
    type=str,
    help='Find all Schedule Stop Pairs from origin on date')
@click.option(
    '--service-from-date',
    required=False,
    default=None,
    type=str,
    help='Find all Schedule Stop Pairs in effect from a date')
@click.option(
    '--service-before-date',
    required=False,
    default=None,
    type=str,
    help='Find all Schedule Stop Pairs in effect before a date')
@click.option(
    '--origin-departure-between',
    required=False,
    default=None,
    type=str,
    help='Find all Schedule Stop Pairs with origin_departure_time in a range')
@click.option(
    '--trip',
    required=False,
    default=None,
    type=str,
    help='Find all Schedule Stop Pairs by trip identifier')
@click.option(
    '--route-onestop-id',
    required=False,
    multiple=True,
    default=None,
    type=str,
    help='Find all Schedule Stop Pairs by route')
@click.option(
    '--operator-onestop-id',
    required=False,
    multiple=True,
    default=None,
    type=str,
    help='Find all Schedule Stop Pairs by operator')
@click.option(
    '--active/--no-active',
    is_flag=True,
    default=True,
    show_default=True,
    help='Schedule Stop Pairs from active FeedVersions')
@click.option(
    '-p',
    '--per-page',
    required=False,
    default=50,
    show_default=True,
    type=int,
    help='Number of results per page')
@click.option(
    '--page-all/--no-page-all',
    is_flag=True,
    default=False,
    show_default=True,
    help='Page over all responses')
def schedule_stop_pairs(**kwargs):
    """Request schedule stop pairs info"""
    kwargs = handle_geometry(**kwargs)
    features_iter = transitland.schedule_stop_pairs(**kwargs)
    write_to_stdout(features_iter)


@click.command()
@click.option(
    '--oid',
    required=False,
    default=None,
    type=str,
    help='a Onestop ID for any type of entity (for example, a stop or an operator)'
)
@click.option(
    '-f',
    '--file',
    required=False,
    default=None,
    type=click.Path(exists=True, file_okay=True, readable=True),
    help='a file with one or more Onestop IDs, with each on their own line.')
def onestop_id(oid, file):
    """Request onestop_id info"""
    if sum(list(map(bool, [oid, file]))) != 1:
        raise ValueError('must provide either oid or file')

    if oid:
        r_iter = transitland.onestop_id(oid)
        for res in r_iter:
            click.echo(json.dumps(res, separators=(',', ':')))

    else:
        with open(file) as f:
            for line in f:
                _id = line.strip()
                r_iter = transitland.onestop_id(oid=_id)
                for res in r_iter:
                    click.echo(json.dumps(res, separators=(',', ':')))


@click.command()
@click.option(
    '-b',
    '--bbox',
    required=False,
    default=None,
    type=str,
    help='Bounding box to search within')
@click.option(
    '-g',
    '--geometry',
    required=False,
    default=None,
    type=click.Path(exists=True, file_okay=True, readable=True),
    help='File with geometry to use. Must be readable by GeoPandas')
@click.option(
    '-p',
    '--per-page',
    required=False,
    default=50,
    show_default=True,
    type=int,
    help='Number of results per page')
@click.option(
    '--page-all/--no-page-all',
    is_flag=True,
    default=False,
    show_default=True,
    help='Page over all responses')
def feeds(**kwargs):
    """Request feeds info"""
    kwargs = handle_geometry(**kwargs)
    features_iter = transitland.feeds(**kwargs)
    write_to_stdout(features_iter)


def handle_geometry(**kwargs):
    bbox = kwargs.pop('bbox')
    geometry_file = kwargs.pop('geometry')

    if sum(list(map(bool, [bbox, geometry_file]))) > 1:
        raise ValueError('must provide either bbox or geometry')

    geometry = None
    if bbox:
        bbox = list(map(float, bbox.split(',')))
        geometry = box(*bbox)
    elif geometry_file:
        geometry = load_file(geometry_file)

    kwargs['geometry'] = geometry
    return kwargs


def load_file(file):
    """Load file into GeoDataFrame
    """
    import geopandas as gpd

    gdf = gpd.read_file(file)
    # Reproject to EPSG 4326
    gdf = gdf.to_crs(epsg=4326)
    # Coalesce into single geometry
    return gdf.unary_union


def write_to_stdout(features_iter):
    """Write features to stdout
    """
    for features in features_iter:
        for feature in features:
            click.echo(json.dumps(feature, separators=(',', ':')))


main.add_command(stops)
main.add_command(operators)
main.add_command(routes)
main.add_command(schedule_stop_pairs)
main.add_command(route_stop_patterns)
main.add_command(onestop_id)
main.add_command(feeds)

if __name__ == '__main__':
    main()
